File: JsonSchemaAndStructs/KiteConnect-Kotlin-Structs/instruments_all.kt
// To parse the JSON, install kotlin's serialization plugin and do:
//
// val json           = Json(JsonConfiguration.Stable)
// val instrumentsAll = json.parse(InstrumentsAll.serializer(), jsonString)

package InstrumentsAll

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*

typealias InstrumentsAll = JsonArray<InstrumentsAllElement>

@Serializable
data class InstrumentsAllElement (
    val exchange: Exchange? = null,

    @SerialName("exchange_token")
    val exchangeToken: Long? = null,

    val expiry: String? = null,

    @SerialName("instrument_token")
    val instrumentToken: Long? = null,

    @SerialName("instrument_type")
    val instrumentType: InstrumentType? = null,

    @SerialName("last_price")
    val lastPrice: Double? = null,

    @SerialName("lot_size")
    val lotSize: Long? = null,

    val name: String? = null,
    val segment: Segment? = null,
    val strike: Long? = null,

    @SerialName("tick_size")
    val tickSize: Double? = null,

    val tradingsymbol: String? = null
)

@Serializable
enum class Exchange(val value: String) {
    Bse("BSE"),
    Nfo("NFO"),
    Nse("NSE");

    companion object : KSerializer<Exchange> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("InstrumentsAll.Exchange", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): Exchange = when (val value = decoder.decodeString()) {
            "BSE" -> Bse
            "NFO" -> Nfo
            "NSE" -> Nse
            else  -> throw IllegalArgumentException("Exchange could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: Exchange) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class InstrumentType(val value: String) {
    Ce("CE"),
    Eq("EQ"),
    PE("PE");

    companion object : KSerializer<InstrumentType> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("InstrumentsAll.InstrumentType", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): InstrumentType = when (val value = decoder.decodeString()) {
            "CE" -> Ce
            "EQ" -> Eq
            "PE" -> PE
            else -> throw IllegalArgumentException("InstrumentType could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: InstrumentType) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class Segment(val value: String) {
    Bse("BSE"),
    NfoOpt("NFO-OPT"),
    Nse("NSE");

    companion object : KSerializer<Segment> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("InstrumentsAll.Segment", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): Segment = when (val value = decoder.decodeString()) {
            "BSE"     -> Bse
            "NFO-OPT" -> NfoOpt
            "NSE"     -> Nse
            else      -> throw IllegalArgumentException("Segment could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: Segment) {
            return encoder.encodeString(value.value)
        }
    }
}

File: JsonSchemaAndStructs/KiteConnect-Rust-Structs/instruments_all.rs
// Example code that deserializes and serializes the model.
// extern crate serde;
// #[macro_use]
// extern crate serde_derive;
// extern crate serde_json;
//
// use generated_module::[object Object];
//
// fn main() {
//     let json = r#"{"answer": 42}"#;
//     let model: [object Object] = serde_json::from_str(&json).unwrap();
// }

extern crate serde_derive;

pub type InstrumentsAll = Vec<InstrumentsAllElement>;

#[derive(Debug, Serialize, Deserialize)]
pub struct InstrumentsAllElement {
    pub exchange: Option<Exchange>,
    pub exchange_token: Option<i64>,
    pub expiry: Option<String>,
    pub instrument_token: Option<i64>,
    pub instrument_type: Option<InstrumentType>,
    pub last_price: Option<f64>,
    pub lot_size: Option<i64>,
    pub name: Option<String>,
    pub segment: Option<Segment>,
    pub strike: Option<i64>,
    pub tick_size: Option<f64>,
    pub tradingsymbol: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Exchange {
    #[serde(rename = "BSE")]
    Bse,
    #[serde(rename = "NFO")]
    Nfo,
    #[serde(rename = "NSE")]
    Nse,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum InstrumentType {
    #[serde(rename = "CE")]
    Ce,
    #[serde(rename = "EQ")]
    Eq,
    #[serde(rename = "PE")]
    Pe,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Segment {
    #[serde(rename = "BSE")]
    Bse,
    #[serde(rename = "NFO-OPT")]
    NfoOpt,
    #[serde(rename = "NSE")]
    Nse,
}
